lib/firestore.js

```diff
@@ -0,0 +1,124 @@
+/* @flow */
+
+import { isString, isNull, isObject, isArray, isNumber, isDate, isBoolean } from './types'
+import { promiseSerial } from './utility'
+
+import fs from 'fs'
+import mkdirp from 'mkdirp'
+
+export const constructReferenceUrl = (reference: Object) => {
+  var referencePath = ''
+  Object.keys(reference).forEach(key => {
+    Object.keys(reference[key]).forEach(subKey => {
+      if (subKey === 'segments') {
+        const pathArray = reference[key][subKey]
+        pathArray.forEach(pathKey => { referencePath = referencePath ? `${referencePath}/${pathKey}` : pathKey })
+      }
+    })
+  })
+  return referencePath ? { value: referencePath, type: 'reference' } : { value: reference, type: 'unknown' }
+}
+
+export const constructDocumentValue = (documentDataToStore: Object = {}, keys: Array<string>, documentData: Object) => {
+  keys.forEach(key => {
+    // Boolean - boolean
+    // Reference - reference
+    // Integer - number
+    // Array - array
+    // Object - object
+    // Float - number
+    // Geographical Point - todo
+    // Map = todo
+    // Null - null
+    // String - string
+    if (isBoolean(documentData[key])) {
+      documentDataToStore = Object.assign({}, documentDataToStore, { [key]: isBoolean(documentData[key]) })
+    } else if (isDate(documentData[key])) {
+      documentDataToStore = Object.assign({}, documentDataToStore, { [key]: isDate(documentData[key]) })
+    } else if (isNumber(documentData[key])) {
+      documentDataToStore = Object.assign({}, documentDataToStore, { [key]: isNumber(documentData[key]) })
+    } else if (isArray(documentData[key])) {
+      documentDataToStore[key] = Object.assign({}, documentDataToStore[key], { type: 'array' })
+      documentDataToStore[key] = Object.assign({}, documentDataToStore[key], constructDocumentValue({}, Object.keys(documentData[key]), documentData[key]))
```

Comment:
Is there a reason for assigning to `documentDataToStore[key]` twice? Would this be synonymous with 
```js
      documentDataToStore[key] = Object.assign({}, documentDataToStore[key], { type: 'array' }, constructDocumentValue({}, Object.keys(documentData[key]), documentData[key]))

```

or is there something that misses?

Reply:
You can do this in one statement also. This code is not optimised. So you can fix this type of code.
I will do optimisation once I am free.